juju/juju
# Run storage hooks for caas models

## Description of change

We now run storage hooks for caas charms with storage. The main logic changes are quite small. Some base storage watcher methods were modified to take a names.Tag instead of MachineTag, since caas storage is associated with units. Much of the content is adding new tests for these watcher changes (testing was a bit incomplete before).

As a driveby, ensure --attach-storage option is not allowed for caas add-unit and deploy

## QA steps

Deploy a caas charm with storage.
run show-status-log and check the storage attached hook has run
